crates/walrus-sdk/src/config/client_config.rs

```diff
@@ -0,0 +1,470 @@
+// Copyright (c) Walrus Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::HashMap,
+    num::{NonZeroU16, NonZeroUsize},
+};
+
+use serde::{Deserialize, Serialize};
+use sui_types::base_types::ObjectID;
+use walrus_core::{
+    encoding::{EncodingConfig, EncodingConfigTrait as _, Primary},
+    EncodingType,
+};
+
+use super::{ClientCommunicationConfig, Config};
+
+/// Multi config for the client.
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+#[serde(untagged)]
+#[allow(clippy::large_enum_variant)]
+pub enum MultiConfig {
+    SingletonConfig(Config),
+    MultiConfig {
+        contexts: HashMap<String, Config>,
+        default_context: String,
+    },
+}
+
+/// Represents one or more exchange objects to be used for SUI/WAL exchange.
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+#[serde(untagged)]
+pub enum ExchangeObjectConfig {
+    /// A single exchange is configured.
+    One(ObjectID),
+    /// Multiple exchanges are configured. A random one is chosen to perform the exchange.
+    Multiple(Vec<ObjectID>),
+}
+
+/// Communication limits in the client.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CommunicationLimits {
+    /// The maximum number of concurrent writes to the storage nodes.
+    pub max_concurrent_writes: usize,
+    /// The maximum number of concurrent sliver reads from the storage nodes.
+    pub max_concurrent_sliver_reads: usize,
+    /// The maximum number of concurrent metadata reads from the storage nodes.
+    pub max_concurrent_metadata_reads: usize,
+    /// The maximum number of concurrent status reads from the storage nodes.
+    pub max_concurrent_status_reads: usize,
+    /// The maximum amount of data in flight (in bytes) for the client.
+    pub max_data_in_flight: usize,
+}
```

Comment:
I have a few suggestions:
- (No strong opinion on this) I feel `MultiConfig` can be placed near [Client]Config in src/config.rs, since it's small and also more of an implementation detail of [Client]Config. We can also remove `pub` after moving there.
- (Not caused by your PR, but since I already see it), ExchangeObjectConfig is no longer used and can be removed.
- (if the first two are adopted) If CommunicationLimits is the only struct left in this file, maybe consider renaming this file to something like src/config/communication_limits.rs?

Reply:
I lumped the CommunicationLimits into communication_config.rs, seems to make sense there. I also renamed MultiConfig to MultiClientConfig, since that's more inline with the rename to ClientConfig. Deleted ExchangeObjectConfig.